src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java

```diff
@@ -237,37 +223,184 @@ public static Object deserializeStoredVector(BytesRef storedVector, VectorDataTy
      * @return expected vector length
      */
     public static int getExpectedVectorLength(final KNNVectorFieldType knnVectorFieldType) {
-        int expectedDimensions = knnVectorFieldType.getDimension();
-        if (isModelBasedIndex(expectedDimensions)) {
+        int expectedDimensions;
+        if (ANNConfigType.FROM_MODEL == knnVectorFieldType.getAnnConfig().getAnnConfigType()) {
```

Comment:
This could be handled by ANNConfig class.
knnVectorFieldType.getAnnConfig().getDimension(knnVectorFieldType);

If we have different class for each AnnConfig, it can handle it properly from its own method and we can avoid this if/else.

Reply:
Let me see if I can make change. The main issue around the model components is that they are not available during node recovery because they are not always added to cluster state. Ill take a closer look to see what I can do to clean this up